### PR TITLE
Ability of user to create registration is fixed

### LINKED ISSUE #829: Allow confirmed speakers to register even after registration period is over
Taken from [this comment](https://github.com/openSUSE/osem/pull/795#issuecomment-190727264).

> We should probably provision for confirmed speakers to be able to register even after the registration period (they will be there anyway).
> All that might collide with registration_limit, we should address it.
> @differentreality 

We need to be more explicit with authorization of registration, `new` and `create` actions should be allowed if user is confirmed speaker.

### LINKED ISSUE #1628: User can only  register once for every conference
https://github.com/openSUSE/osem/blob/master/app/models/ability.rb#L69 Ability of user to register should not return true when he have already register.

### LINKED ISSUE #1827: Speaker proposals allow registration outside registration period
**I'm submitting a ..**
- [X] Bug Report
- [ ] Feature Request

**Current behavior:**
After submitting a proposal (Event), a number of steps are presented to the user, to finish their proposal. One of those steps is event registration, regardless of whether the registration period is open for the conference.

**Expected behavior:**
No registration link is presented outside of registration period; authorization blocks registration outside of registration period for non-admin users.

**Steps to reproduce:**

1. Create a conference; do not set a registration period.
2. Create a Call for Papers, type: events.
3. Submit a proposal.
4. Perform the "Registration" task.

**Related code:**

**Other information:**

### PR DESCRIPTION
**Pull request by namangupta01**
Closes #1628 
Closes #1827 
Closes #829

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@
 ActiveRecord::Schema.define(version: 20170721001700) do
 
   create_table "ahoy_events", force: :cascade do |t|
-    t.integer  "visit_id"
+    t.uuid     "visit_id",   limit: 16
```
Comment (Ana06) [inline db/schema.rb]: this change is unrelated, isn't it? :confused:

File: app/models/ability.rb [hunk 2]
```diff
@@ -66,7 +66,7 @@ def signed_in(user)
 
     can [:new, :create], Registration do |registration|
       conference = registration.conference
-      conference.registration_open? && !conference.registration_limit_exceeded? || conference.program.speakers.confirmed.include?(user)
+      (conference.registration_open? && !conference.registration_limit_exceeded? || conference.program.speakers.confirmed.include?(user))&&!(conference.user_registered? user)
```
Comment (Ana06) [inline app/models/ability.rb]: This line got too complicate to read.

what about writing it like:

``` Ruby
return false if conference.user_registered? user
return true if conference.program.speakers.confirmed.include?(user)
conference.registration_open? && !conference.registration_limit_exceeded?
```

File: db/schema.rb (hunk 1 above)
Comment (namangupta01) [inline db/schema.rb]: This get automatically changed when i run the command rake db:migrate

File: app/models/ability.rb (hunk 2 above)
Comment (namangupta01) [inline app/models/ability.rb]: Thanks :) This looks good !!

File: db/schema.rb (hunk 1 above)
Comment (Ana06) [inline db/schema.rb]: @namangupta01 it seems it comes from here: https://github.com/openSUSE/osem/pull/1229

@differentreality @eug3nix could be a migration missed?

@namangupta01 anyway it is unrelated to your changes, so please exclude it from your PR. :wink:

**Comment by namangupta01** [comment]
@Ana06  Is it now good to go ?

**Comment by namangupta01** [comment]
@Ana06 @differentreality can anyone one of you help me to understand why my Travis build is failing? Thanks:)
https://travis-ci.org/openSUSE/osem/builds/268890801?utm_source=github_status&utm_medium=notification

**Comment by differentreality** [comment]
@namangupta01 sure! How much do you understand? And which part do you not understand?

**Comment by namangupta01** [comment]
@differentreality Thank you :) I understand that the error is occurring because of the line https://github.com/namangupta01/osem/blob/e065079b6bb116780c5d06895b545299af29048e/app/models/ability.rb#L70
But i didn't understand why the error is saying **unexpected return** as my syntax is correct.
https://github.com/namangupta01/osem/blob/e065079b6bb116780c5d06895b545299af29048e/app/models/ability.rb#L70

**Comment by differentreality** [comment]
@namangupta01 you said your syntax is correct, yet there is an error. 
Did you google the error to figure out what is going on?

**Comment by namangupta01** [comment]
@differentreality Oh! I got why it was failing. Thanks :)

File: app/models/ability.rb [hunk 3]
```diff
@@ -66,7 +66,13 @@ def signed_in(user)
 
     can [:new, :create], Registration do |registration|
       conference = registration.conference
-      conference.registration_open? && !conference.registration_limit_exceeded? || conference.program.speakers.confirmed.include?(user)
+      if conference.user_registered? user
+        false
```
Comment (differentreality) [inline app/models/ability.rb]: Would it make sense to add a `cannot` block for this case after this can block?

Comment (namangupta01) [reply app/models/ability.rb]: @differentreality I didn't get what case are you taking about? But as far as i think you are taking about  **if conference.user_registered? user**.
I donot think why we should add a cannot block for this?

Comment (namangupta01) [reply app/models/ability.rb]: As this can block is used to check wheather a user can register or not, so i think all the cases that depends on wheather a user can register or not should be in same block.

**Comment by namangupta01** [comment]
@differentreality @Ana06  Please review this pull request

**Comment by differentreality** [comment]
We should enhance our tests accordingly.

**Comment by differentreality** [comment]
Great @namangupta01
Can you please show me where we test the case that the user is a speaker?

**Comment by namangupta01** [comment]
@differentreality still stuck in that example :(

File: spec/models/ability_spec.rb [hunk 4]
```diff
@@ -95,6 +95,51 @@
...
+          it do
+            event_with_speaker_confirmed.speakers << user
+            should be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed))
+          end
+        end
+
+        context 'when registration are closed and registration_limit_exceeded' do
+          let(:conference_with_closed_registration_and_limit_excedeed) { create(:conference, registration_limit: '1') }
+          let!(:closed_registration_period) { create(:registration_period, conference: conference_with_closed_registration_and_limit_excedeed, start_date: Date.current - 6.days, end_date: Date.current - 6.days) }
+          let(:program_with_speaker_confirmed) { create(:program, conference: conference_with_closed_registration_and_limit_excedeed) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Also not needed

File: spec/models/ability_spec.rb [hunk 5]
```diff
@@ -95,6 +95,51 @@
       it{ should_not be_able_to(:new, Registration.new(conference_id: conference_with_closed_registration.id))}
       it{ should_not be_able_to(:create, Registration.new(conference_id: conference_with_closed_registration.id))}
 
+      # Test for user can register or not
+      let(:registration_for_registering_user) { create(:registration, conference: create(:conference), user: user) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Perhaps it is easier to understand what we are testing with this, if it is called user_registration ?

File: spec/models/ability_spec.rb [hunk 6]
```diff
@@ -95,6 +95,51 @@
...
+      it{ should_not be_able_to(:new, Registration.new(conference_id: registration_for_registering_user.conference_id)) }
+      it{ should_not be_able_to(:create, Registration.new(conference_id: registration_for_registering_user.conference_id)) }
+
+      context 'when user is a speaker' do
+        context 'when speaker is confirmed' do
+          let(:conference_with_speaker_confirmed) { create(:conference) }
+          let(:program_with_speaker_confirmed) { create(:program, conference: conference_with_speaker_confirmed) }
+          let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: program_with_speaker_confirmed) }
+
+          it do
```
[the remark this answers is not in the document]
Comment (namangupta01) [reply spec/models/ability_spec.rb]: All the things are defined in the context what can we give more in 'it block' description?

[the remark this answers is not in the document]
Comment (namangupta01) [reply spec/models/ability_spec.rb]: Does this description is correct?

screen shot 2017-11-16 at 2 15 18 pm

**Comment by namangupta01** [comment]
@differentreality Please have a look!

**Comment by namangupta01** [comment]
@differentreality Is this fine now? Should i squash the commits now?

**Comment by bear454** [comment]
@namangupta01: check out https://github.com/openSUSE/osem/pull/1834/commits/dae9cea34d7f531ecf9ae03e79a9506a7256e028

File: app/models/ability.rb [hunk 7]
```diff
@@ -66,7 +66,13 @@ def signed_in(user)
...
+      elsif conference.program.speakers.confirmed.include? user
```
Comment (bear454) [inline app/models/ability.rb]:
```
elsif (conference.program.speakers.confirmed.include?(user) && conference.registration_period)
  # Confirmed speakers can register before or after a registration period, but there _must be_ a registration period.
```

**Comment by namangupta01** [comment]
@bear454 respective changes are done and tests are implemented accordingly.

**Comment by bear454** [comment]
@differentreality, @Ana06 : re-evaluate reviews, please?

File: spec/models/ability_spec.rb [hunk 8]
```diff
@@ -95,6 +95,67 @@
...
+          end
+
+          it 'checks that speaker with event confirmed with registration period set can create registration' do
+            event_with_speaker_confirmed.speakers << user
+            conference_with_speaker_confirmed.registration_period = create(:registration_period)
+            should be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed))
+          end
+        end
+
+        context 'when registration are closed and registration_limit_exceeded' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: This is not one single test case. I would expect that we test at least for the following:
- registration open + limit_exceeded 
- registration closed + limit **not** exceeded

Comment (namangupta01) [reply spec/models/ability_spec.rb]: @differentreality Will correct it 👍

File: spec/models/ability_spec.rb [hunk 9]
```diff
@@ -95,6 +95,67 @@
...
+            should_not be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed))
+          end
+
+          it 'checks that speaker with event confirmed with registration period not set cannot create registration' do
+            event_with_speaker_confirmed.speakers << user
+            should_not be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed))
+          end
+        end
+
+        context 'when event is confirmed and the registration period is set' do
```
[the remark this answers is not in the document]
Comment (namangupta01) [reply spec/models/ability_spec.rb]: registration period open ? Speaker can do registration only when the registration period is set regardless of the whether registration is open or closed.

[the remark this answers is not in the document]
Comment (namangupta01) [reply spec/models/ability_spec.rb]: I didn't get the point of using open here and also in other comments.

File: spec/models/ability_spec.rb [hunk 10]
```diff
@@ -95,6 +95,67 @@
...
+          let(:conference_with_speaker_confirmed) { create(:conference) }
+          let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
+
+          it 'checks that speaker with event confirmed with registration period set can open new registration' do
```
[the remark this answers is not in the document]
Comment (namangupta01) [reply spec/models/ability_spec.rb]: Shouldn't it be `when user is a speaker with event confirmed and registration period is set the speaker can access #new action for Registration` ?

**Comment by differentreality** [comment]
@namangupta01 from the comments it seems we are not on the same page on which cases we need to test for. Can you please make a table with all the necessary cases? Preferably numbered, so that we can confirm which ones we are currently covering in our tests, and which ones we aren't.

**Comment by namangupta01** [comment]
@differentreality Please look at the cases! I will implement it once you verify it!

|  | User Registered? | Registration open? | Speaker's event is confirmed (if user is speaker) | Registration Set | Conference Limit Exceed?| User can register?
|---|------------------|------------------|--------------------|-----|-----|---|
|1.|no|no|no (USER OR SPEAKER)|no|no|no
|2.|no|no|no (USER OR SPEAKER)|yes|no|no|
|3.|no|no|no (USER OR SPEAKER)|yes|yes|no|
|4.|no|yes|no (USER OR SPEAKER)|yes|yes|no|
|5.|no|yes|no (USER OR SPEAKER)|yes|no|yes|
|6.|yes|-|-|-|-|-|no|
|7.|no|no|yes|yes|yes|yes|
|8.|no|no|yes|yes|no|yes|
|9.|no|no|yes|no|no|no|
|10.|no|yes|yes|yes|yes|yes|
|11.|no|yes|yes|yes|yes|yes|
|12.|no|yes|yes|yes|no|yes|

**Comment by namangupta01** [comment]
Please let me know if i left some cases! And cases 1 to 3 can be implemented in one case do i have to implement it in different cases.

File: spec/models/ability_spec.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (differentreality) [reply spec/models/ability_spec.rb]: So 1 case is No9 from your array (registration_period NOT set being the important part).
If the registration_period is set, what we care about is whether or not it is open.
When open, obviously the speaker of a confirmed event, and anyone else, can register.
When not open, the speaker of a confirmed event can still register. <-- That's the important case here

[the remark this answers is not in the document]
Comment (namangupta01) [reply spec/models/ability_spec.rb]: Yes you are right! will update it accordingly!

**Comment by namangupta01** [comment]
@differentreality All tests are implemented. Have a look!

**Comment by bear454** [comment]
Finite state machines FTW :)

File: spec/models/ability_spec.rb [hunk 11]
```diff
@@ -89,11 +89,102 @@
...
+
+      end
+
+      context 'when user is a speaker with event confirmed' do
+        context 'when the registration period is not set' do
+          let(:conference_with_speaker_confirmed) { create(:conference) }
+          let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
+
+          it 'cannot access #new action for Registration' do
+            event_with_speaker_confirmed.speakers << user
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Move this to a before block?

File: spec/models/ability_spec.rb [hunk 12]
```diff
@@ -89,11 +89,102 @@
...
+        end
+
+        context 'when user have already registered' do
+          let(:user_registration) { create(:registration, conference: conference, user: user) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: user_registration.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: user_registration.conference_id)) }
+        end
+
+        context 'when user have not registered and registrations are open with conference_limit_not_exceeded' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: The hidden not in conference_limit_not_exceeded is not very obvious. 
`registration_limit_not_exceeded` derives from the model function that checks this https://github.com/openSUSE/osem/blob/master/app/models/conference.rb#L712, changing this too much, and adding text makes it harder to read in my opinion. 
This can benefit from a small re-write.

File: spec/models/ability_spec.rb [hunk 13]
```diff
@@ -89,11 +89,102 @@
...
+          let(:registration_period) { create(:registration_period) }
+          let(:conference_with_open_registration_and_limit_not_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
+
+          it{ should be_able_to(:new, Registration.new(conference: conference_with_open_registration_and_limit_not_excedeed)) }
+          it{ should be_able_to(:create, Registration.new(conference: conference_with_open_registration_and_limit_not_excedeed)) }
+        end
+
+        context 'when user have not registered and registrations are open with conference_limit_exceeded' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: _has_ not

Comment (differentreality) [inline spec/models/ability_spec.rb]: it's registration limit, not conference. We should either use the function name as it is, or use free text.

File: spec/models/ability_spec.rb [hunk 14]
```diff
@@ -89,11 +89,102 @@
...
+          let(:registration_period) { create(:registration_period) }
+          let(:conference_with_open_registration_and_limit_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
+          let(:registration_for_user_with_open_registration_and_limit_excedeed) { create(:registration, conference: conference_with_open_registration_and_limit_excedeed, user: user2) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+        end
+
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: remove empty line

File: spec/models/ability_spec.rb [hunk 15]
```diff
@@ -89,11 +89,102 @@
...
+        context 'when the registration period is not set' do
+          let(:conference_with_speaker_confirmed) { create(:conference) }
+          let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
+
+          it 'cannot access #new action for Registration' do
+            event_with_speaker_confirmed.speakers << user
+            should_not be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed))
+          end
+
+          it 'can access #create action for Registration' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: can or cannot? You test for should_not

File: spec/models/ability_spec.rb [hunk 16]
```diff
@@ -89,11 +89,102 @@
...
+            it 'can access #new action for Registration' do
+              event_with_speaker_confirmed.speakers << user
+              should be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed))
+            end
+
+            it 'can access #create action for Registration' do
+              event_with_speaker_confirmed.speakers << user
+              should be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed))
+            end
+
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Unnecessary empty line

**Comment by namangupta01** [comment]
@differentreality I didn't take the case when user is confirmed speaker, registration is open and registration limit exceeded because when the user is confirmed speaker and registration is open then it does not depend whether registration limit exceeded or not, user can always register in this case.

**Comment by differentreality** [comment]
It is supposed to be able to always register. Tests are there to test whether or not our code works as expected.

**Comment by namangupta01** [comment]
ok i will implement this test also

**Comment by namangupta01** [comment]
@differentreality Test for the Case when user is a confirmed speaker, registration is open, and the registration limit has exceeded? has been added. Have a look!

**Comment by namangupta01** [comment]
@differentreality Please review!

File: spec/models/ability_spec.rb [hunk 17]
```diff
@@ -89,11 +89,103 @@
...
+          it{ should_not be_able_to(:new, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+        end
+      end
+
+      context 'when user is a speaker with event confirmed' do
+        context 'when the registration period is not set' do
+          let(:conference_with_speaker_confirmed) { create(:conference) }
+          let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: You are defining this in multiple places, it seems it should be defined once after line 136

File: spec/models/ability_spec.rb [hunk 18]
```diff
@@ -89,11 +89,103 @@
...
+
+          before :each do
+            event_with_speaker_confirmed.speakers << user
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Then you can also extract this before block after line 136

File: spec/models/ability_spec.rb [hunk 19]
```diff
@@ -89,11 +89,103 @@
...
+
+            before :each do
+              event_with_speaker_confirmed.speakers << user
+            end
+
+            it{ should be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed)) }
+            it{ should be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed)) }
+
+            context 'and registration_limit_exceeded' do
+              let(:conference_with_open_registration_and_limit_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Based on the previous comments, instead of this let statement, perhaps you can just change the registration_limit in the existing conference?

File: spec/models/ability_spec.rb [hunk 20]
```diff
@@ -89,11 +89,103 @@
...
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_closed_registration)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_closed_registration)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference)) }
+        end
+
+        context 'when user has already registered' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Let's also test this for when the user is a speaker with confirmed event.

**Comment by namangupta01** [comment]
@differentreality please re-review!

File: spec/models/ability_spec.rb [hunk 21]
```diff
@@ -89,11 +89,104 @@
...
-      it{ should_not be_able_to(:new, Registration.new(conference_id: conference_with_closed_registration.id))}
-      it{ should_not be_able_to(:create, Registration.new(conference_id: conference_with_closed_registration.id))}
+
+      # Test for user can register or not
+      context 'when user is not a speaker with event confirmed' do
+        let(:conference) { create(:conference) }
+        let(:registration_period) { create(:registration_period) }
+
+        context 'when the registration is closed' do
+          let(:conference_with_closed_registration) { create(:conference) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: What's the difference between conference and conference_with_closed_registration? They are both created by the same factory, yet you expect one to have a registration which is over, and the other one to not have a registration_period at all.

File: spec/models/ability_spec.rb [hunk 22]
```diff
@@ -89,11 +89,104 @@
...
+
+        context 'when user has already registered' do
+          let(:user_registration) { create(:registration, conference: conference, user: user) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: user_registration.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: user_registration.conference_id)) }
+        end
+
+        context 'when user has not registered and registrations are open with no registration_limit_exceeded' do
+          let(:conference_with_open_registration_and_limit_not_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: This way of associating the newly created conference with the registration_period is not very good, as it is not very explicit which conference the registration_period should belong to.

Why do you need to create a new conference here? Isn't it enough that we have a conference (or should have) with a registration period set? Then you can manipulate the dates of that registration_period and register your user in that conference.

File: spec/models/ability_spec.rb [hunk 23]
```diff
@@ -89,11 +89,104 @@
...
+      context 'when user is a speaker with event confirmed' do
+        let(:conference_with_speaker_confirmed) { create(:conference) }
+        let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
+
+        before :each do
+          event_with_speaker_confirmed.speakers << user
+        end
+
+        context 'when speaker has already registered' do
+
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Please remove this empty line

File: spec/models/ability_spec.rb [hunk 24]
```diff
@@ -89,11 +89,104 @@
...
+      context 'when user is a speaker with event confirmed' do
+        let(:conference_with_speaker_confirmed) { create(:conference) }
+        let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
+
+        before :each do
+          event_with_speaker_confirmed.speakers << user
+        end
+
+        context 'when speaker has already registered' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: For this to be tested properly, you need to have a conference with an existing registration_period. If there is no registration period this test would be passing not because the already registered user cannot register, but because the registration period is not set. 

While writing your tests, imagine you have not written your code yet. Also your code can change at any time, the order of the checks can be different or totally altered. The tests need to be independent, and run successfully no matter the code. For that reason, you need to set them up properly

File: spec/models/ability_spec.rb [hunk 25]
```diff
@@ -89,11 +89,104 @@
...
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed)) }
+        end
+
+        context 'And the registration period is set' do
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: What about we keep some uniformity on the context text?

File: spec/models/ability_spec.rb [hunk 26]
```diff
@@ -89,11 +89,104 @@
...
+          let(:registration_period) { create(:registration_period) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: In these tests, I would always expect the registration_period to be associated with a relevant conference immediately. It is a bit confusing how you later on associate it with a different conference.

Think about it logically, what do you do first if you are a user in OSEM?
Don't you first create a new conference, and then go setup a registration_period (for that specific conference)?
If you keep this logical order, it is way easier to read the tests later on.

This line can simply be
create(:registration_period, conference: conference_with_speaker_confirmed)

File: spec/models/ability_spec.rb [hunk 27]
```diff
@@ -89,11 +89,104 @@
...
+
+          before :each do
+            conference_with_speaker_confirmed.registration_period = registration_period
+          end
+
+          context 'with registration open' do
+
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Please remove empty line (there other relevant places for this to happen)

File: spec/models/ability_spec.rb [hunk 28]
```diff
@@ -89,11 +89,104 @@
...
+                create(:registration, conference: conference_with_speaker_confirmed, user: user2)
+              end
+
+              it{ should be_able_to(:new, Registration.new(conference: conference_with_speaker_confirmed)) }
+              it{ should be_able_to(:create, Registration.new(conference: conference_with_speaker_confirmed)) }
+            end
+          end
+
+          context 'with registration closed' do
+            let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_closed_registration.program) }
```
Comment (differentreality) [inline spec/models/ability_spec.rb]: Where is it defined that this conference has a closed registration?

File: spec/models/ability_spec.rb (hunk 21 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: Yes, you are right about that they both created form the same factory but there is a difference.
https://github.com/openSUSE/osem/blob/master/spec/models/ability_spec.rb#L33
That's how conference and conference_with_closed_registration are different.
The only thing is that i shouldn't have create the conference_with_closed_registration here as it is already created in https://github.com/openSUSE/osem/blob/master/spec/models/ability_spec.rb#L32

File: spec/models/ability_spec.rb (hunk 22 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: Yes we can do that! The only thing that i was trying to do is that it would make easier to understand the code!

File: spec/models/ability_spec.rb (hunk 23 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: okay 👍

File: spec/models/ability_spec.rb (hunk 28 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: check this https://github.com/openSUSE/osem/blob/master/spec/models/ability_spec.rb#L33

File: spec/models/ability_spec.rb (hunk 22 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: One of the main thing about let statement is that the value will be cached
across multiple calls in the same example but not across examples. That's why i don't think it's a problem. Please correct me if i am wrong!

File: spec/models/ability_spec.rb (hunk 24 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: Yes you are right about that tests are passing because registration period is not set.

File: spec/models/ability_spec.rb (hunk 26 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: I associate a conference with a different conference because value will be cached
across multiple calls in the same example but not across examples. So when there is another example test it won't be affect by the previous or any other test.

File: spec/models/ability_spec.rb (hunk 20 above)
Comment (namangupta01) [reply spec/models/ability_spec.rb]: This test is already there.

**Comment by namangupta01** [comment]
Hey @differentreality I have done the changes please have a look.

**Comment by namangupta01** [comment]
@differentreality Please re-review!

**Comment by namangupta01** [comment]
@bear454 Can you please see this?

**Comment by bear454** [comment]
@namangupta01 I re-triggered the test suite; it's green now, but I can't change @Ana06 or @differentreality 's reviews.

**Comment by differentreality** [comment]
@namangupta01 :+1:  please squash your commits

**Comment by namangupta01** [comment]
sure!

**Comment by namangupta01** [comment]
@differentreality Done squashing!

**Comment by namangupta01** [comment]
@differentreality something needs to be change or it is good to go?

**Comment by differentreality** [comment]
Sorry for the delay @namangupta01 I have been exceptionally busy the past few weeks. 

Thanks so much for working on this feature :-) Very good work :+1:
